listenbrainz/tests/integration/test_api.py

```diff
@@ -133,16 +133,25 @@ def test_get_listens_with_time_range(self):
         self.assert200(response)
         self.assertEqual(response.json['status'], 'ok')
 
-        # This sleep allows for the timescale subscriber to take its time in getting
-        # the listen submitted from redis and writing it to timescale.
-        # Removing it causes an empty list of listens to be returned.
-        time.sleep(2)
-
+        count = 0
         url = url_for('api_v1.get_listens', user_name=user['musicbrainz_id'])
-        response = self.client.get(url)
+        expected_count = 3
+        while count < 5:
+            count += 1
+            # This sleep allows for the timescale subscriber to take its time in getting
+            # the listen submitted from redis and writing it to timescale.
+            # Removing it causes an empty list of listens to be returned.
+            # We try a few times (up to a maximum of 5) to see if the data is present,
+            # but then fail if the data still hasn't arrived
+            time.sleep(2)
+
+            response = self.client.get(url)
+            data = json.loads(response.data)['payload']
+            if data['count'] == expected_count:
```

Comment:
Is there a possibility that this would result in a false positive, for example if I'm expecting 3 listens but in the end 4 listens arrive, will this test just pass as soon as it gets 3, missing the erroneous 4th?

Reply:
2 things to prevent this from happening:
- we only submit 3 items (the `for i in range(3):....` above)
- we only use this check to break out of the sleep loop quickly. A few lines down we still have an assert to check that the count value is expected. If there are 4, the assert on l154 will fail.

Reply:
Right, thanks :)